### ISSUE TITLE
[react-redux] DefaultProps support

### ISSUE THREAD

**Issue by fabiomcosta** [flow-typed/flow-typed#3137]
I was trying to add `connect` to a component that has `defaultProps` and I had a hard time getting it right, even thought I could.

One thing I noticed while looking for ways to do it, is that there is an issue on the current tests: https://github.com/flow-typed/flow-typed/blob/master/definitions/npm/react-redux_v5.x.x/flow_v0.89.x-/test_connect.js#L10

That line should be `passthroughWithDefaultProp: number,` (not optional) because that component has `passthroughWithDefaultProp` as a defaultProp.
When trying to use `passthroughWithDefaultProp` inside that component, even though that prop is always a number (guaranteed by defaultProp) Flow will complain that it can be a number or undefined, which is not ideal.

Also note that on this Flow documentation example `foo` is declared as `number` and not an options `number: https://flow.org/en/docs/react/hoc/#toc-exporting-wrapped-components

This try-flow link shows the problem I'm trying to point out: https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBLAtgBzgJwBdkwBDAZzACUBTUgY2KnzizAHJ87H2BudYMDABJAHb0CXRgC4wAV3IZRAczBwchDHFGkYYKATJgAJjShKaxk2dJyYhAAosc5dIQCeOGmCfryARjAAXjAAbwAfVDAwQTAAFQALDEpyBLg7K1E4YgAjb3VNbV0wPPpbcnyoMEIE71NzUUtrKFt7XxcAOiiwAA8AfllyQnwlZVRwgF9+VHoYCkoADTAaHsIaUWNKWgZCDoBhVjxG0UIAHnaAgD4w7qHSTXpm1sdnShDQ7uie2XZ2bonulwNjR8AAKACUN2i0ViIJY+E+YC4hDk+FE1SS5A6OFeHR6YAA1Bw+P9UADULEDvgpIRZAA5bLeNJIABi8CQCG07CYpAwegwVQQ3g8XjAGGIFDAAAN3INhqMpWAEOKEhi6jY7C8-G5PN4LgAmYJhSLRWVgIYjFTjKboWbzMAATWWq3Wm2o3F2B1w2nWZwN1w+0TuDyemouRsD0LNv1J0SBpjBkMjMKEWWW1IIAEJEcjUeiaslsbj3ITifxogDybEGWswMywGzEEquTy+XWQTQADQlOSt+xK2rorhYOAAN1G0vqzwuitBFW8jVWztIuBgNHBqFBpydwEusm2jA6AFE11hfacRTQ4FUHZdwdNYgApBTEVKICf1xtIFp8yjKmpGFk6IFFoOh6Di6hKiq6TEFKU5hq8Uo6qKFwAMwRiaYAAF5ypaYyTNMdrkJQABazprBsWwevshw+ic5yvKhAaAq6IIQlC0K5miGKFhBnRYaWvzlmAlYCEIR4ZgiW5kbu+7USeNBnvRl7XmAJF3rwQA

Here is how I made it work, but maybe there is a way to avoid having to use `React.Config` ourselves?:

```js
import * as React from 'react';

type OwnProps = {|
  x: number
|};

type Props = {|
  ...OwnProps,
  // no state props
|};

class Comp extends React.Component<Props> {
  static defaultProps = { x: 3 };
  ...
}

type DefaultProps = {| x: number |};
type Config = React.Config<Props, DefaultProps>;
type OwnConfig = React.Config<OwnProps, DefaultProps>;

connect<Config, OwnConfig, _, _, _, _>()(Comp);
```

**Comment by jbrown215** [comment]
In the way Flow models react, `Props` should be the type of `this.props` and `Config` is the type of the Props with all defaultProps marked optional. The way you handled it in your example at the bottom of this issue is exactly correct!

**Comment by AndrewSouthpaw** [comment]
Nice work @fabiomcosta, would you mind documenting this somewhere useful? Maybe in the react-redux libdef?

**Comment by fabiomcosta** [comment]
@AndrewSouthpaw I'm guessing changing the unit tests to use `Config` would also be good.
I'll try to find time this week.

**Comment by mike1808** [comment]
I agree, that the user shouldn't use `React.Config` and the type def of `react-redux` should use it.

**Comment by retyui** [comment]
I found a solution for this problem without critical changes, I just ran into another problem.

We can get `defaultProps` from `WrappedComponent` and then Use React.Config for create `ConnectedComponentProps`

[Try ...](https://flow.org/try/#0PTAEAEDMBsHsHcBQiAuBPADgU1AES5AIYCu0KACgE6wYDOoAvKAN4A+oAboZQJaEB2KAFyhaKXvwDmoVgF8A3MnTZQAeXj8qNekzaJQoAHTH8RUhWp0ANPtA8AJiP7EAtgCMslG3MWpMOLTpGFiNjdU1LWitQFB4UaCwAMWoXAFlCDABlFEIULAAVWEDaETEJaQVkAGNoQlp6AHVKDOx7AGFYFwxYfixBUCwADzz+e3oAJSxCKpQAEg6unr6UAB5igD4WWxBQACEAVXzQeAALPtAqzu7e-vsegHIUUBPCDhxQAAMxXJ4q0HsCCQyMUPsccJIsE9YCgzpQBpRqJRtmA2gJ+NC7PxvoI+HlPrMAKIJFzLfL+UEeKokWg4DCWTzoT4AszAyKgnj0Fwc2g8KSYgzfWJVeiwSCfJotLDtK5LQQfZACnJC-6A8zFESmIEWbTBZicbh8QQie41KaUe6gSqyZB+FQdfi9GZShbXZbFYKTaZze2QHiSFbhYrRQnE0n+FbKLCi0ASjCtF2yp7Re7MrXFe7rdY2rhwlAARhE9sdeWlixu2qCumAACoc4bhKB7vcrNXgA4RHmrJVEDmYgAmQs9YvOmXl926fW8AQNpvRdugPOW3w7QmDbAzAkI2BwuvT0AuYhiUAeUTiXmSHvcGIAZkHDqwTtLrsE45Cu6NC7njgXS+QK4Ja4Pigm6InY9h2PQlBYAAjsQPBQeBdI0JeuYAAx3sOT6Jq+ervjOzY1mBHYtsAlRAA)
```js
// @flow

type DefaultProps = {| variant: string |};

type OwnProps = {|
  ...DefaultProps,
  id: number,
|};

type Props = { ...OwnProps, titleFromMapStateToProps: string };

class WrappedComponent extends React$Component<Props> {
  // BUT when component don't have  `static defaultProps` we get other error
  // Cannot instantiate `$ElementType` because property `defaultProps` is missing in  statics of `WrappedComponent`

  static defaultProps: DefaultProps = { variant: 'clear' };
}


type ConnectedComponentProps = React$Config<OwnProps, $ElementType<typeof WrappedComponent , 'defaultProps'>>


var t1: ConnectedComponentProps = {/*variant: '',*/id: 1,};

var t2: ConnectedComponentProps = { variant: '', id: 1 };

// $ExpectError variant must be string
var t3: ConnectedComponentProps = { variant: 1, id: 1 };

// $ExpectError id is required prop
var t0: ConnectedComponentProps = { variant: '',/* id: 1,*/};
```
